From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aloweeed <134882723+aloweeed@users.noreply.github.com>
Date: Sat, 8 Aug 2026 03:40:39 +0300
Subject: [PATCH] bro.


diff --git a/src/main/java/dev/florium/FloriumEntityScheduler.java b/src/main/java/dev/florium/FloriumEntityScheduler.java
new file mode 100644
index 0000000000000000000000000000000000000000..6bf44096d02542ca771321d1a06a1880f54917bc
--- /dev/null
+++ b/src/main/java/dev/florium/FloriumEntityScheduler.java
@@ -0,0 +1,173 @@
+package dev.florium;
+
+import ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.profiling.ProfilerFiller;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.decoration.ArmorStand;
+import net.minecraft.world.entity.item.FallingBlockEntity;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.chunk.LevelChunk;
+
+public final class FloriumEntityScheduler {
+
+    // Florium start - hard optimize lag entities
+
+    private static final int CHUNK_THRESHOLD = 40;
+
+    private static final int TARGET_TICKS_PER_CHUNK = 200;
+
+    private static final int MAX_INTERVAL = 20;
+
+    private final ServerLevel level;
+
+    public FloriumEntityScheduler(final ServerLevel level) {
+        this.level = level;
+    }
+
+    public void tick(
+            final IteratorSafeOrderedReferenceSet.Iterator<Entity> iterator,
+            final ProfilerFiller profiler
+    ) {
+        try {
+            final long gameTime = this.level.getGameTime();
+
+            while (iterator.hasNext()) {
+                final Entity entity = iterator.next();
+
+                if (!shouldTick(entity, gameTime)) {
+                    handleSkippedTick(entity);
+                    continue;
+                }
+
+                beforeRealTick(entity);
+                tickEntity(entity, profiler);
+            }
+        } finally {
+            iterator.finishedIterating();
+        }
+    }
+
+    private boolean shouldTick(
+            final Entity entity,
+            final long gameTime
+    ) {
+        if (!isLagEntity(entity)) {
+            return true;
+        }
+
+        if (entity.isRemoved()) {
+            return false;
+        }
+
+        final ChunkPos pos = entity.chunkPosition();
+
+        final LevelChunk chunk =
+                this.level.getChunkSource().getChunkNow(pos.x(), pos.z());
+
+        if (chunk == null) {
+            return true;
+        }
+
+        final int count = chunk.florium$lagEntitiesCounter;
+
+        final int CHUNK_THRESHOLD = (entity instanceof ArmorStand) ? 15 : 40;
+        final int TARGET_TICKS_PER_CHUNK = (entity instanceof ArmorStand) ? 15 : 80;
+        final int MAX_INTERVAL = (entity instanceof ArmorStand) ? 40 : 20;
+
+        if (count <= CHUNK_THRESHOLD) {
+            return true;
+        }
+        final int interval = Math.clamp(
+                ceilDiv(count, TARGET_TICKS_PER_CHUNK),
+                1,
+                MAX_INTERVAL
+        );
+
+        if (interval == 1) {
+            return true;
+        }
+
+        final int entityPhase =
+                Math.floorMod(entity.getId(), interval);
+
+        final int currentPhase =
+                Math.floorMod(gameTime, interval);
+
+        return entityPhase == currentPhase;
+    }
+
+    private static void handleSkippedTick(final Entity entity) {
+        if (entity instanceof FallingBlockEntity fallingBlock) {
+            ++fallingBlock.florium$skippedTicks;
+        }
+    }
+
+    private static void beforeRealTick(final Entity entity) {
+        if (entity instanceof FallingBlockEntity fallingBlock) {
+            final int skipped = fallingBlock.florium$skippedTicks;
+
+            if (skipped != 0) {
+                fallingBlock.time += skipped;
+                fallingBlock.florium$skippedTicks = 0;
+            }
+        }
+    }
+
+    private static boolean isLagEntity(final Entity entity) {
+        return entity instanceof ArmorStand
+                || entity instanceof FallingBlockEntity;
+    }
+
+    private static int ceilDiv(
+            final int value,
+            final int divisor
+    ) {
+        return (value + divisor - 1) / divisor;
+    }
+
+    private void tickEntity(
+            final Entity entity,
+            final ProfilerFiller profiler
+    ) {
+        if (entity.isRemoved()) {
+            return;
+        }
+
+        if (this.level.tickRateManager().isEntityFrozen(entity)) {
+            return;
+        }
+
+        profiler.push("checkDespawn");
+        entity.checkDespawn();
+        profiler.pop();
+
+//        if (entity instanceof ArmorStand) return;
+
+        // Folia - region threading
+        if (entity.isRemoved()) {
+            return;
+        }
+
+        final Entity vehicle = entity.getVehicle();
+
+        if (vehicle != null) {
+            if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
+                return;
+            }
+
+            entity.stopRiding();
+        }
+
+        profiler.push("tick");
+
+        this.level.guardEntityTick(
+                this.level::tickNonPassenger,
+                entity
+        );
+
+        profiler.pop();
+    }
+
+    // Florium end - hard optimize lag entities
+}
\ No newline at end of file
diff --git a/src/main/java/dev/florium/FloriumUtil.java b/src/main/java/dev/florium/FloriumUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..383df299512ea015117544aba6a1f03b76e3d103
--- /dev/null
+++ b/src/main/java/dev/florium/FloriumUtil.java
@@ -0,0 +1,117 @@
+package dev.florium;
+
+import ca.spottedleaf.common.time.TickData;
+import io.papermc.paper.adventure.PaperAdventure;
+import io.papermc.paper.threadedregions.ThreadedRegionizer;
+import io.papermc.paper.threadedregions.TickRegionScheduler;
+import io.papermc.paper.threadedregions.TickRegions;
+import io.papermc.paper.threadedregions.scheduler.FoliaAsyncScheduler;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.ComponentLike;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
+import net.kyori.adventure.text.minimessage.tag.standard.StandardTags;
+import net.minecraft.network.chat.TextColor;
+import net.minecraft.network.protocol.game.ClientboundBossEventPacket;
+import net.minecraft.resources.Identifier;
+import net.minecraft.server.bossevents.CustomBossEvent;
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.Bukkit;
+import org.bukkit.boss.BossBar;
+import org.bukkit.craftbukkit.boss.CraftKeyedBossbar;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.entity.Player;
+
+import java.awt.*;
+import java.text.DecimalFormat;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+public class FloriumUtil {
+    public static final MiniMessage MM = MiniMessage.miniMessage();
+    public static final String GRADIENT_TEMPLATE = "<gradient:#8EF56A:#37D67A:#00C2A8>";
+    public static final net.kyori.adventure.text.format.TextColor COLOR_A = net.kyori.adventure.text.format.TextColor.color(0xFF2222);
+    public static final net.kyori.adventure.text.format.TextColor COLOR_B = net.kyori.adventure.text.format.TextColor.color(0xAAFFAA);
+    public static final Component FLORIUM_COLORED = formatGradient("Florium");
+
+    private static long LAST_UPDATE = 0;
+    private static final long INTERVAL = 2000; // 2 sec.
+    private static final CustomBossEvent BAR = new CustomBossEvent(UUID.randomUUID(), Identifier.withDefaultNamespace("florium-bossbar"), net.minecraft.network.chat.Component.literal("test"), () -> {});
+
+    private static final Executor ASYNC = Executors.newSingleThreadScheduledExecutor();
+    private static final ThreadLocal<DecimalFormat> TWO_DECIMAL_PLACES = ThreadLocal.withInitial(() -> {
+        return new DecimalFormat("#,##0.00");
+    });
+
+    private static final Set<ServerPlayer> STAT_LISTENERS = ConcurrentHashMap.newKeySet();
+    public static boolean toggleStatListener(Player player) {
+        ServerPlayer serverPlayer = ((CraftPlayer) player).getHandle();
+        if (STAT_LISTENERS.contains(serverPlayer)) {
+            serverPlayer.connection.send(ClientboundBossEventPacket.createRemovePacket(BAR.getId()));
+            STAT_LISTENERS.remove(serverPlayer);
+            return false;
+        } else {
+            serverPlayer.connection.send(ClientboundBossEventPacket.createAddPacket(BAR));
+            STAT_LISTENERS.add(serverPlayer);
+            return true;
+        }
+    }
+    public static void updateStat() {
+        if (System.currentTimeMillis() - LAST_UPDATE > INTERVAL) {
+            final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
+                    TickRegionScheduler.getCurrentRegion();
+            final long currTime = System.nanoTime();
+
+            final TickData.TickReportData report15s = region.getData().getRegionSchedulingHandle().getTickReport15s(currTime);
+            final double tps15s = report15s.tpsData().segmentAll().average();
+            final double mspt15s = report15s.timePerTickData().segmentAll().average() / 1.0E6;
+
+            final double tps_factor = tps15s / 20f;
+            final double mspt_factor = 1f - (mspt15s / 100f);
+            Runtime runtime = Runtime.getRuntime();
+            final long used = (runtime.maxMemory() - runtime.freeMemory());
+            final double mb = 1024.0 * 1024.0 * 1024.0;
+            final double ram = used / mb;
+
+            Component format = Component.text("TPS: ").color(NamedTextColor.GRAY)
+                    .append(transition(TWO_DECIMAL_PLACES.get().format(tps15s), tps_factor))
+                    .append(Component.text(" MSPT: ").color(NamedTextColor.GRAY))
+                    .append(transition(TWO_DECIMAL_PLACES.get().format(mspt15s), mspt_factor))
+                    .append(Component.text(" RAM: ").color(NamedTextColor.GRAY))
+                    .append(transition(TWO_DECIMAL_PLACES.get().format(ram), 1f - ((double) used / runtime.maxMemory())))
+                    .append(Component.text("/").color(NamedTextColor.GRAY))
+                    .append(transition(TWO_DECIMAL_PLACES.get().format(runtime.maxMemory() / mb), 1f));
+            BAR.setName(PaperAdventure.asVanilla(format));
+            BAR.setProgress((float) (tps_factor * mspt_factor));
+            LAST_UPDATE = System.currentTimeMillis();
+            ClientboundBossEventPacket updateName = ClientboundBossEventPacket.createUpdateNamePacket(BAR);
+            ClientboundBossEventPacket updateProgress = ClientboundBossEventPacket.createUpdateProgressPacket(BAR);
+            for (ServerPlayer player : STAT_LISTENERS) {
+                player.connection.send(updateName);
+                player.connection.send(updateProgress);
+            }
+        }
+    }
+
+    public static Component formatGradient(String m) {
+        return MM.deserialize(GRADIENT_TEMPLATE.concat(m).concat("</gradient>"));
+    }
+    public static String formatGradientString(String m) {
+        return GRADIENT_TEMPLATE.concat(m).concat("</gradient>");
+    }
+    public static Component transition(String m, double state) {
+        return Component.text(m).color(net.kyori.adventure.text.format.TextColor.lerp((float) Math.clamp(state, 0f, 1f), COLOR_A, COLOR_B));
+    }
+
+    public static ComponentLike getVersion() {
+        return FLORIUM_COLORED.append(MM.deserialize(" running on <gray>".concat(Bukkit.getVersion())));
+    }
+
+    public static void runAsync(Runnable task) {
+        ASYNC.execute(task);
+    }
+}
diff --git a/src/main/java/dev/florium/commands/FloriumStatCommand.java b/src/main/java/dev/florium/commands/FloriumStatCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..b4d45038bf111f9979a3e563eb256346e00de19e
--- /dev/null
+++ b/src/main/java/dev/florium/commands/FloriumStatCommand.java
@@ -0,0 +1,41 @@
+package dev.florium.commands;
+
+import com.mojang.brigadier.Command;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.tree.LiteralCommandNode;
+import dev.florium.FloriumUtil;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import io.papermc.paper.command.brigadier.Commands;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Player;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class FloriumStatCommand {
+
+    private static final Component ONLY_PLAYERS = Component.text("Only players can execute this command").color(NamedTextColor.RED);
+    private static final Component ENABLED = Component.text("Stat bar enabled").color(NamedTextColor.GREEN);
+    private static final Component DISABLED = Component.text("Stat bar disabled").color(NamedTextColor.RED);
+
+    public static LiteralCommandNode<CommandSourceStack> create() {
+        final FloriumStatCommand command = new FloriumStatCommand();
+        return Commands.literal("stat")
+            .requires(source -> source.getSender().hasPermission("florium.command.stat"))
+            .executes(command::execute)
+            .build();
+    }
+
+    private int execute(CommandContext<CommandSourceStack> context) {
+        CommandSender sender = context.getSource().getSender();
+        if (sender instanceof Player player) {
+            sender.sendMessage(
+                    FloriumUtil.toggleStatListener(player) ? ENABLED : DISABLED
+            );
+        } else {
+            sender.sendMessage(ONLY_PLAYERS);
+        }
+        return Command.SINGLE_SUCCESS;
+    }
+}
