From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aloweeed <134882723+aloweeed@users.noreply.github.com>
Date: Sat, 8 Aug 2026 04:18:55 +0300
Subject: [PATCH] apply all


diff --git a/dev/florium/FloriumEntityScheduler.java b/dev/florium/FloriumEntityScheduler.java
new file mode 100644
index 0000000000000000000000000000000000000000..4fb6b2d6c6a61d81484250d91a108cce2df7f244
--- /dev/null
+++ b/dev/florium/FloriumEntityScheduler.java
@@ -0,0 +1,173 @@
+package dev.florium;
+
+import ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.profiling.ProfilerFiller;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.decoration.ArmorStand;
+import net.minecraft.world.entity.item.FallingBlockEntity;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.chunk.LevelChunk;
+
+public final class FloriumEntityScheduler {
+
+    // Florium start - hard optimize lag entities
+
+    private static final int CHUNK_THRESHOLD = 40;
+
+    private static final int TARGET_TICKS_PER_CHUNK = 200;
+
+    private static final int MAX_INTERVAL = 20;
+
+    private final ServerLevel level;
+
+    public FloriumEntityScheduler(final ServerLevel level) {
+        this.level = level;
+    }
+
+    public void tick(
+            final IteratorSafeOrderedReferenceSet.Iterator<Entity> iterator,
+            final ProfilerFiller profiler
+    ) {
+        try {
+            final long gameTime = this.level.getGameTime();
+
+            while (iterator.hasNext()) {
+                final Entity entity = iterator.next();
+
+                if (!shouldTick(entity, gameTime)) {
+                    handleSkippedTick(entity);
+                    continue;
+                }
+
+                beforeRealTick(entity);
+                tickEntity(entity, profiler);
+            }
+        } finally {
+            iterator.finishedIterating();
+        }
+    }
+
+    private boolean shouldTick(
+            final Entity entity,
+            final long gameTime
+    ) {
+        if (!isLagEntity(entity)) {
+            return true;
+        }
+
+        if (entity.isRemoved()) {
+            return false;
+        }
+
+        final ChunkPos pos = entity.chunkPosition();
+
+        final LevelChunk chunk =
+                this.level.getChunkSource().getChunkNow(pos.x(), pos.z());
+
+        if (chunk == null) {
+            return true;
+        }
+
+        final int count = chunk.florium$lagEntities;
+
+        final int CHUNK_THRESHOLD = (entity instanceof ArmorStand) ? 15 : 40;
+        final int TARGET_TICKS_PER_CHUNK = (entity instanceof ArmorStand) ? 15 : 80;
+        final int MAX_INTERVAL = (entity instanceof ArmorStand) ? 40 : 20;
+
+        if (count <= CHUNK_THRESHOLD) {
+            return true;
+        }
+        final int interval = Math.clamp(
+                ceilDiv(count, TARGET_TICKS_PER_CHUNK),
+                1,
+                MAX_INTERVAL
+        );
+
+        if (interval == 1) {
+            return true;
+        }
+
+        final int entityPhase =
+                Math.floorMod(entity.getId(), interval);
+
+        final int currentPhase =
+                Math.floorMod(gameTime, interval);
+
+        return entityPhase == currentPhase;
+    }
+
+    private static void handleSkippedTick(final Entity entity) {
+        if (entity instanceof FallingBlockEntity fallingBlock) {
+            ++fallingBlock.florium$skippedTicks;
+        }
+    }
+
+    private static void beforeRealTick(final Entity entity) {
+        if (entity instanceof FallingBlockEntity fallingBlock) {
+            final int skipped = fallingBlock.florium$skippedTicks;
+
+            if (skipped != 0) {
+                fallingBlock.time += skipped;
+                fallingBlock.florium$skippedTicks = 0;
+            }
+        }
+    }
+
+    public static boolean isLagEntity(final Entity entity) {
+        return entity instanceof ArmorStand
+                || entity instanceof FallingBlockEntity;
+    }
+
+    private static int ceilDiv(
+            final int value,
+            final int divisor
+    ) {
+        return (value + divisor - 1) / divisor;
+    }
+
+    private void tickEntity(
+            final Entity entity,
+            final ProfilerFiller profiler
+    ) {
+        if (entity.isRemoved()) {
+            return;
+        }
+
+        if (this.level.tickRateManager().isEntityFrozen(entity)) {
+            return;
+        }
+
+        profiler.push("checkDespawn");
+        entity.checkDespawn();
+        profiler.pop();
+
+//        if (entity instanceof ArmorStand) return;
+
+        // Folia - region threading
+        if (entity.isRemoved()) {
+            return;
+        }
+
+        final Entity vehicle = entity.getVehicle();
+
+        if (vehicle != null) {
+            if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
+                return;
+            }
+
+            entity.stopRiding();
+        }
+
+        profiler.push("tick");
+
+        this.level.guardEntityTick(
+                this.level::tickNonPassenger,
+                entity
+        );
+
+        profiler.pop();
+    }
+
+    // Florium end - hard optimize lag entities
+}
\ No newline at end of file
diff --git a/dev/florium/FloriumUtil.java b/dev/florium/FloriumUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..383df299512ea015117544aba6a1f03b76e3d103
--- /dev/null
+++ b/dev/florium/FloriumUtil.java
@@ -0,0 +1,117 @@
+package dev.florium;
+
+import ca.spottedleaf.common.time.TickData;
+import io.papermc.paper.adventure.PaperAdventure;
+import io.papermc.paper.threadedregions.ThreadedRegionizer;
+import io.papermc.paper.threadedregions.TickRegionScheduler;
+import io.papermc.paper.threadedregions.TickRegions;
+import io.papermc.paper.threadedregions.scheduler.FoliaAsyncScheduler;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.ComponentLike;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
+import net.kyori.adventure.text.minimessage.tag.standard.StandardTags;
+import net.minecraft.network.chat.TextColor;
+import net.minecraft.network.protocol.game.ClientboundBossEventPacket;
+import net.minecraft.resources.Identifier;
+import net.minecraft.server.bossevents.CustomBossEvent;
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.Bukkit;
+import org.bukkit.boss.BossBar;
+import org.bukkit.craftbukkit.boss.CraftKeyedBossbar;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.entity.Player;
+
+import java.awt.*;
+import java.text.DecimalFormat;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+public class FloriumUtil {
+    public static final MiniMessage MM = MiniMessage.miniMessage();
+    public static final String GRADIENT_TEMPLATE = "<gradient:#8EF56A:#37D67A:#00C2A8>";
+    public static final net.kyori.adventure.text.format.TextColor COLOR_A = net.kyori.adventure.text.format.TextColor.color(0xFF2222);
+    public static final net.kyori.adventure.text.format.TextColor COLOR_B = net.kyori.adventure.text.format.TextColor.color(0xAAFFAA);
+    public static final Component FLORIUM_COLORED = formatGradient("Florium");
+
+    private static long LAST_UPDATE = 0;
+    private static final long INTERVAL = 2000; // 2 sec.
+    private static final CustomBossEvent BAR = new CustomBossEvent(UUID.randomUUID(), Identifier.withDefaultNamespace("florium-bossbar"), net.minecraft.network.chat.Component.literal("test"), () -> {});
+
+    private static final Executor ASYNC = Executors.newSingleThreadScheduledExecutor();
+    private static final ThreadLocal<DecimalFormat> TWO_DECIMAL_PLACES = ThreadLocal.withInitial(() -> {
+        return new DecimalFormat("#,##0.00");
+    });
+
+    private static final Set<ServerPlayer> STAT_LISTENERS = ConcurrentHashMap.newKeySet();
+    public static boolean toggleStatListener(Player player) {
+        ServerPlayer serverPlayer = ((CraftPlayer) player).getHandle();
+        if (STAT_LISTENERS.contains(serverPlayer)) {
+            serverPlayer.connection.send(ClientboundBossEventPacket.createRemovePacket(BAR.getId()));
+            STAT_LISTENERS.remove(serverPlayer);
+            return false;
+        } else {
+            serverPlayer.connection.send(ClientboundBossEventPacket.createAddPacket(BAR));
+            STAT_LISTENERS.add(serverPlayer);
+            return true;
+        }
+    }
+    public static void updateStat() {
+        if (System.currentTimeMillis() - LAST_UPDATE > INTERVAL) {
+            final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
+                    TickRegionScheduler.getCurrentRegion();
+            final long currTime = System.nanoTime();
+
+            final TickData.TickReportData report15s = region.getData().getRegionSchedulingHandle().getTickReport15s(currTime);
+            final double tps15s = report15s.tpsData().segmentAll().average();
+            final double mspt15s = report15s.timePerTickData().segmentAll().average() / 1.0E6;
+
+            final double tps_factor = tps15s / 20f;
+            final double mspt_factor = 1f - (mspt15s / 100f);
+            Runtime runtime = Runtime.getRuntime();
+            final long used = (runtime.maxMemory() - runtime.freeMemory());
+            final double mb = 1024.0 * 1024.0 * 1024.0;
+            final double ram = used / mb;
+
+            Component format = Component.text("TPS: ").color(NamedTextColor.GRAY)
+                    .append(transition(TWO_DECIMAL_PLACES.get().format(tps15s), tps_factor))
+                    .append(Component.text(" MSPT: ").color(NamedTextColor.GRAY))
+                    .append(transition(TWO_DECIMAL_PLACES.get().format(mspt15s), mspt_factor))
+                    .append(Component.text(" RAM: ").color(NamedTextColor.GRAY))
+                    .append(transition(TWO_DECIMAL_PLACES.get().format(ram), 1f - ((double) used / runtime.maxMemory())))
+                    .append(Component.text("/").color(NamedTextColor.GRAY))
+                    .append(transition(TWO_DECIMAL_PLACES.get().format(runtime.maxMemory() / mb), 1f));
+            BAR.setName(PaperAdventure.asVanilla(format));
+            BAR.setProgress((float) (tps_factor * mspt_factor));
+            LAST_UPDATE = System.currentTimeMillis();
+            ClientboundBossEventPacket updateName = ClientboundBossEventPacket.createUpdateNamePacket(BAR);
+            ClientboundBossEventPacket updateProgress = ClientboundBossEventPacket.createUpdateProgressPacket(BAR);
+            for (ServerPlayer player : STAT_LISTENERS) {
+                player.connection.send(updateName);
+                player.connection.send(updateProgress);
+            }
+        }
+    }
+
+    public static Component formatGradient(String m) {
+        return MM.deserialize(GRADIENT_TEMPLATE.concat(m).concat("</gradient>"));
+    }
+    public static String formatGradientString(String m) {
+        return GRADIENT_TEMPLATE.concat(m).concat("</gradient>");
+    }
+    public static Component transition(String m, double state) {
+        return Component.text(m).color(net.kyori.adventure.text.format.TextColor.lerp((float) Math.clamp(state, 0f, 1f), COLOR_A, COLOR_B));
+    }
+
+    public static ComponentLike getVersion() {
+        return FLORIUM_COLORED.append(MM.deserialize(" running on <gray>".concat(Bukkit.getVersion())));
+    }
+
+    public static void runAsync(Runnable task) {
+        ASYNC.execute(task);
+    }
+}
diff --git a/dev/florium/commands/FloriumStatCommand.java b/dev/florium/commands/FloriumStatCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..b4d45038bf111f9979a3e563eb256346e00de19e
--- /dev/null
+++ b/dev/florium/commands/FloriumStatCommand.java
@@ -0,0 +1,41 @@
+package dev.florium.commands;
+
+import com.mojang.brigadier.Command;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.tree.LiteralCommandNode;
+import dev.florium.FloriumUtil;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import io.papermc.paper.command.brigadier.Commands;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Player;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class FloriumStatCommand {
+
+    private static final Component ONLY_PLAYERS = Component.text("Only players can execute this command").color(NamedTextColor.RED);
+    private static final Component ENABLED = Component.text("Stat bar enabled").color(NamedTextColor.GREEN);
+    private static final Component DISABLED = Component.text("Stat bar disabled").color(NamedTextColor.RED);
+
+    public static LiteralCommandNode<CommandSourceStack> create() {
+        final FloriumStatCommand command = new FloriumStatCommand();
+        return Commands.literal("stat")
+            .requires(source -> source.getSender().hasPermission("florium.command.stat"))
+            .executes(command::execute)
+            .build();
+    }
+
+    private int execute(CommandContext<CommandSourceStack> context) {
+        CommandSender sender = context.getSource().getSender();
+        if (sender instanceof Player player) {
+            sender.sendMessage(
+                    FloriumUtil.toggleStatListener(player) ? ENABLED : DISABLED
+            );
+        } else {
+            sender.sendMessage(ONLY_PLAYERS);
+        }
+        return Command.SINGLE_SUCCESS;
+    }
+}
diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index cb15cfbc684ae9e9d9634886f707c2df10063139..5668a9ba21452d3470ee688ee5442f004da6b5c3 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -594,15 +594,8 @@ public final class RegionizedWorldData {
         TickRegions.RegionStats.updateCurrentRegion();
     }
 
-    public void forEachTickingEntity(final Consumer<Entity> action) {
-        final IteratorSafeOrderedReferenceSet.Iterator<Entity> iterator = this.entityTickList.iterator();
-        try {
-            while (iterator.hasNext()) {
-                action.accept(iterator.next());
-            }
-        } finally {
-            iterator.finishedIterating();
-        }
+    public IteratorSafeOrderedReferenceSet.Iterator<Entity> getTickEntities() { // Florium
+        return this.entityTickList.iterator();
     }
 
     public void addEntity(final Entity entity) {
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 6f13870a1b463049dd37cf1c65d0bdecfc35b5c4..d6471cb23bc51617e5d14287198b84c8e97483cf 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -10,6 +10,7 @@ import com.mojang.datafixers.DataFixer;
 import com.mojang.jtracy.DiscontinuousFrame;
 import com.mojang.jtracy.TracyClient;
 import com.mojang.logging.LogUtils;
+import dev.florium.FloriumUtil;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectArraySet;
 import java.io.File;
@@ -1784,6 +1785,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper end - Server Tick Events
         this.server.spark.tickEnd(((double)(endTime - startTime) / 1000000D)); // Paper - spark // Folia - region threading
         // Folia - region threading
+
+        FloriumUtil.updateStat();
     }
 
     protected void processPacketsAndTick(final boolean sprinting) {
@@ -1963,7 +1966,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         profiler.popPush("server gui refresh");
 
-        if (false) for (Runnable tickable : this.tickables) { // Folia - region threading - TODO WTF is this?
+        if (false) for (Runnable tickable : this.tickables) { // Folia - region threading - TODO WTF is this? <- um bro??
             tickable.run();
         }
 
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index c6434ff2e5c211887c2e98c8462a49e0e1cb5b21..84c620c373e145eaea7e709b7b0daec8189c6753 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -5,6 +5,7 @@ import com.google.common.collect.Lists;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
+import dev.florium.FloriumEntityScheduler;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.longs.LongSet;
@@ -329,6 +330,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private long tickedBlocksOrFluids;
     // Folia - region threading - move to regionized data
 
+    // Florium start
+    public final FloriumEntityScheduler entityScheduler = new FloriumEntityScheduler(this);
+    // Florium end
+
     @Override
     public final LevelChunk moonrise$getFullChunkIfLoaded(final int chunkX, final int chunkZ) {
         return this.chunkSource.getChunkNow(chunkX, chunkZ);
@@ -398,11 +403,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                                                final ca.spottedleaf.concurrentutil.util.Priority priority,
                                                final java.util.function.Consumer<java.util.List<net.minecraft.world.level.chunk.ChunkAccess>> onLoad) {
         this.moonrise$loadChunksAsync(
-            (pos.getX() - radiusBlocks) >> 4,
-            (pos.getX() + radiusBlocks) >> 4,
-            (pos.getZ() - radiusBlocks) >> 4,
-            (pos.getZ() + radiusBlocks) >> 4,
-            priority, onLoad
+                (pos.getX() - radiusBlocks) >> 4,
+                (pos.getX() + radiusBlocks) >> 4,
+                (pos.getZ() - radiusBlocks) >> 4,
+                (pos.getZ() + radiusBlocks) >> 4,
+                priority, onLoad
         );
     }
 
@@ -411,11 +416,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                                                final net.minecraft.world.level.chunk.status.ChunkStatus chunkStatus, final ca.spottedleaf.concurrentutil.util.Priority priority,
                                                final java.util.function.Consumer<java.util.List<net.minecraft.world.level.chunk.ChunkAccess>> onLoad) {
         this.moonrise$loadChunksAsync(
-            (pos.getX() - radiusBlocks) >> 4,
-            (pos.getX() + radiusBlocks) >> 4,
-            (pos.getZ() - radiusBlocks) >> 4,
-            (pos.getZ() + radiusBlocks) >> 4,
-            chunkStatus, priority, onLoad
+                (pos.getX() - radiusBlocks) >> 4,
+                (pos.getX() + radiusBlocks) >> 4,
+                (pos.getZ() - radiusBlocks) >> 4,
+                (pos.getZ() + radiusBlocks) >> 4,
+                chunkStatus, priority, onLoad
         );
     }
 
@@ -475,7 +480,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         for (int cx = minChunkX; cx <= maxChunkX; ++cx) {
             for (int cz = minChunkZ; cz <= maxChunkZ; ++cz) {
                 ca.spottedleaf.moonrise.common.PlatformHooks.get().scheduleChunkLoad(
-                    this, cx, cz, net.minecraft.world.level.chunk.status.ChunkStatus.FULL, true, priority, consumer
+                        this, cx, cz, net.minecraft.world.level.chunk.status.ChunkStatus.FULL, true, priority, consumer
                 );
             }
         }
@@ -531,12 +536,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         return true;
     }
 
-   @Override
-   public final void moonrise$issueEmergencySave() {
-       this.moonrise$getChunkTaskScheduler().chunkHolderManager.saveAllChunks(
-           true, true, true, true
-       );
-   }
+    @Override
+    public final void moonrise$issueEmergencySave() {
+        this.moonrise$getChunkTaskScheduler().chunkHolderManager.saveAllChunks(
+                true, true, true, true
+        );
+    }
     // Paper end - rewrite chunk system
     // Paper start - chunk tick iteration
     // Folia - region threading
@@ -568,24 +573,24 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     // Paper end - chunk tick iteration
 
     public ServerLevel(
-        final MinecraftServer server,
-        final Executor executor,
-        final LevelStorageSource.LevelStorageAccess levelStorage,
-        final WorldGenSettings worldGenSettings, // CraftBukkit
-        final ResourceKey<Level> dimension,
-        final LevelStem levelStem,
-        final boolean isDebug,
-        final long biomeZoomSeed,
-        final List<CustomSpawner> customSpawners,
-        final boolean tickTime
-        // Paper start - add parameters
-        , ResourceKey<LevelStem> typeKey,
-        org.bukkit.World.Environment env,
-        org.bukkit.generator.ChunkGenerator gen,
-        org.bukkit.generator.BiomeProvider biomeProvider,
-        SavedDataStorage savedDataStorage,
-        io.papermc.paper.world.PaperWorldLoader.LoadedWorldData loadedWorldData
-        // Paper end - add parameters
+            final MinecraftServer server,
+            final Executor executor,
+            final LevelStorageSource.LevelStorageAccess levelStorage,
+            final WorldGenSettings worldGenSettings, // CraftBukkit
+            final ResourceKey<Level> dimension,
+            final LevelStem levelStem,
+            final boolean isDebug,
+            final long biomeZoomSeed,
+            final List<CustomSpawner> customSpawners,
+            final boolean tickTime
+            // Paper start - add parameters
+            , ResourceKey<LevelStem> typeKey,
+            org.bukkit.World.Environment env,
+            org.bukkit.generator.ChunkGenerator gen,
+            org.bukkit.generator.BiomeProvider biomeProvider,
+            SavedDataStorage savedDataStorage,
+            io.papermc.paper.world.PaperWorldLoader.LoadedWorldData loadedWorldData
+            // Paper end - add parameters
     ) {
         // CraftBukkit start
         final io.papermc.paper.world.saveddata.PaperLevelOverrides levelData = loadedWorldData.levelOverrides();
@@ -637,18 +642,18 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         DataFixer fixerUpper = server.getFixerUpper();
         // Paper - rewrite chunk system
         this.chunkSource = new ServerChunkCache(
-            this,
-            levelStorage,
-            fixerUpper,
-            server.getStructureManager(),
-            executor,
-            generator,
-            this.spigotConfig.viewDistance, // Spigot
-            this.spigotConfig.simulationDistance, // Spigot
-            syncWrites,
-            null, // Paper - rewrite chunk system
-            () -> server.overworld().getDataStorage()
-            , savedDataStorage // Paper - initialize SavedDataStorage earlier
+                this,
+                levelStorage,
+                fixerUpper,
+                server.getStructureManager(),
+                executor,
+                generator,
+                this.spigotConfig.viewDistance, // Spigot
+                this.spigotConfig.simulationDistance, // Spigot
+                syncWrites,
+                null, // Paper - rewrite chunk system
+                () -> server.overworld().getDataStorage()
+                , savedDataStorage // Paper - initialize SavedDataStorage earlier
         );
         this.chunkSource.getGeneratorState().ensureStructuresGenerated();
         this.portalForcer = new PortalForcer(this);
@@ -664,17 +669,17 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         WorldOptions options = worldGenSettings.options();
         long seed = options.seed();
         this.structureCheck = new StructureCheck(
-            getTypeKey(), // Paper - pass level stem
-            this.chunkSource.chunkScanner(),
-            this.registryAccess(),
-            server.getStructureManager(),
-            dimension,
-            generator,
-            this.chunkSource.randomState(),
-            this,
-            generator.getBiomeSource(),
-            seed,
-            fixerUpper
+                getTypeKey(), // Paper - pass level stem
+                this.chunkSource.chunkScanner(),
+                this.registryAccess(),
+                server.getStructureManager(),
+                dimension,
+                generator,
+                this.chunkSource.randomState(),
+                this,
+                generator.getBiomeSource(),
+                seed,
+                fixerUpper
         );
         this.structureManager = new StructureManager(this, options, this.structureCheck);
         if (this.dimensionType().hasEnderDragonFight()) {
@@ -691,12 +696,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.moonrise$setEntityLookup(new ca.spottedleaf.moonrise.patches.chunk_system.level.entity.server.ServerEntityLookup((ServerLevel)(Object)this, ((ServerLevel)(Object)this).new EntityCallbacks()));
         this.chunkTaskScheduler = new ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler((ServerLevel)(Object)this);
         this.entityDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.EntityDataController(
-            new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.EntityDataController.EntityRegionFileStorage(
-                new RegionStorageInfo(levelStorage.getLevelId(), dimension, "entities"),
-                levelStorage.getDimensionPath(dimension).resolve("entities"),
-                server.forceSynchronousWrites()
-            ),
-            this.chunkTaskScheduler
+                new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.EntityDataController.EntityRegionFileStorage(
+                        new RegionStorageInfo(levelStorage.getLevelId(), dimension, "entities"),
+                        levelStorage.getDimensionPath(dimension).resolve("entities"),
+                        server.forceSynchronousWrites()
+                ),
+                this.chunkTaskScheduler
         );
         this.poiDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.PoiDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
         this.chunkDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.ChunkDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
@@ -853,11 +858,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             long tick = this.getRedstoneGameTime(); // Folia - region threading
             profiler.push("blockTicks");
             foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.BLOCK_TICK); try { // Folia - profiler
-            regionizedWorldData.getBlockLevelTicks().tick(tick, paperConfig().environment.maxBlockTicks, this::tickBlock); // Paper - configurable max block ticks // Folia - region threading
+                regionizedWorldData.getBlockLevelTicks().tick(tick, paperConfig().environment.maxBlockTicks, this::tickBlock); // Paper - configurable max block ticks // Folia - region threading
             } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.BLOCK_TICK); } // Folia - profiler
             profiler.popPush("fluidTicks");
             foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.FLUID_TICK); try { // Folia - profiler
-            regionizedWorldData.getFluidLevelTicks().tick(tick, paperConfig().environment.maxFluidTicks, this::tickFluid); // Paper - configurable max fluid ticks // Folia - region threading
+                regionizedWorldData.getFluidLevelTicks().tick(tick, paperConfig().environment.maxFluidTicks, this::tickFluid); // Paper - configurable max fluid ticks // Folia - region threading
             } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.FLUID_TICK); } // Folia - profiler
             profiler.pop();
         }
@@ -865,18 +870,18 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         profiler.popPush("raid");
         if (runs) {
             foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.RAIDS_TICK); try { // Folia - profiler
-            this.raids.tick(this);
+                this.raids.tick(this);
             } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.RAIDS_TICK); } // Folia - profiler
         }
 
         profiler.popPush("chunkSource");
         foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.CHUNK_PROVIDER_TICK); try { // Folia - profiler
-        this.getChunkSource().tick(haveTime, true);
+            this.getChunkSource().tick(haveTime, true);
         } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.CHUNK_PROVIDER_TICK); } // Folia - profiler
         profiler.popPush("blockEvents");
         if (runs) {
             foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.BLOCK_EVENT_TICK); try { // Folia - profiler
-            this.runBlockEvents();
+                this.runBlockEvents();
             } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.BLOCK_EVENT_TICK); } // Folia - profiler
         }
 
@@ -895,59 +900,33 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             profiler.push("entities");
             if (this.dragonFight != null && runs) {
                 foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.DRAGON_FIGHT_TICK); try { // Folia - profiler
-                if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this, this.dragonFight.origin)) { // Folia - region threading
-                profiler.push("dragonFight");
-                this.dragonFight.tick();
-                profiler.pop();
-                } else { // Folia start - region threading
-                    // try to load dragon fight
-                    ChunkPos fightCenter = ChunkPos.containing(this.dragonFight.origin);
-                    this.moonrise$getChunkTaskScheduler().chunkHolderManager.addTicketAtLevel(
-                            TicketType.UNKNOWN, fightCenter, ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManager.MAX_TICKET_LEVEL,
-                            null
-                    );
-                } // Folia end - region threading
+                    if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this, this.dragonFight.origin)) { // Folia - region threading
+                        profiler.push("dragonFight");
+                        this.dragonFight.tick();
+                        profiler.pop();
+                    } else { // Folia start - region threading
+                        // try to load dragon fight
+                        ChunkPos fightCenter = ChunkPos.containing(this.dragonFight.origin);
+                        this.moonrise$getChunkTaskScheduler().chunkHolderManager.addTicketAtLevel(
+                                TicketType.UNKNOWN, fightCenter, ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManager.MAX_TICKET_LEVEL,
+                                null
+                        );
+                    } // Folia end - region threading
                 } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.DRAGON_FIGHT_TICK); } // Folia - profiler
             }
 
             foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.ACTIVATE_ENTITIES); try { // Folia - profiler
-            io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
+                io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
             } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.ACTIVATE_ENTITIES); } // Folia - profiler
             foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.ENTITY_TICK); try { // Folia - profiler
-            regionizedWorldData // Folia - regionised ticking
-                .forEachTickingEntity( // Folia - regionised ticking
-                    entity -> {
-                        if (!entity.isRemoved()) {
-                            if (!tickRateManager.isEntityFrozen(entity)) {
-                                profiler.push("checkDespawn");
-                                entity.checkDespawn();
-                                profiler.pop();
-                                if (entity.isRemoved()) return; // Folia - region threading - if we despawned, DON'T TICK IT!
-                                if (true) { // Paper - rewrite chunk system
-                                    Entity vehicle = entity.getVehicle();
-                                    if (vehicle != null) {
-                                        if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
-                                            return;
-                                        }
-
-                                        entity.stopRiding();
-                                    }
-
-                                    profiler.push("tick");
-                                    this.guardEntityTick(this::tickNonPassenger, entity);
-                                    profiler.pop();
-                                }
-                            }
-                        }
-                    }
-                );
+                entityScheduler.tick(regionizedWorldData.getTickEntities(), profiler);
             } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.ENTITY_TICK); } // Folia - profiler
             if (this.paperConfig().unsupportedSettings.ticking.blockEntities) { // Paper - option to disable ticking
-            profiler.popPush("blockEntities");
-            foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.TILE_ENTITY); try { // Folia - profiler
-            this.tickBlockEntities();
-            } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.TILE_ENTITY); } // Folia - profiler
-            profiler.pop();
+                profiler.popPush("blockEntities");
+                foliaProfiler.startTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.TILE_ENTITY); try { // Folia - profiler
+                    this.tickBlockEntities();
+                } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.TILE_ENTITY); } // Folia - profiler
+                profiler.pop();
             } // Paper - option to disable ticking
             regionizedWorldData.currentPrimedTnt = 0; // Spigot // Folia - regionised ticking
         }
@@ -958,7 +937,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         profiler.push("debugSynchronizers");
         if (this.debugSynchronizers.hasAnySubscriberFor(DebugSubscriptions.NEIGHBOR_UPDATES)) {
             regionizedWorldData.neighborUpdater // Folia - region threading
-                .setDebugListener(blockPos -> this.debugSynchronizers.broadcastEventToTracking(blockPos, DebugSubscriptions.NEIGHBOR_UPDATES, blockPos));
+                    .setDebugListener(blockPos -> this.debugSynchronizers.broadcastEventToTracking(blockPos, DebugSubscriptions.NEIGHBOR_UPDATES, blockPos));
         } else {
             regionizedWorldData.neighborUpdater.setDebugListener(null); // Folia - region threading
         }
@@ -1020,7 +999,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         final ca.spottedleaf.leafprofiler.RegionizedProfiler.Handle profiler = io.papermc.paper.threadedregions.TickRegionScheduler.getProfiler(); // Folia - profiler
         for (CustomSpawner spawner : this.customSpawners) {
             final int customSpawnerTimer = profiler.getOrCreateTimerAndStart(() -> "Misc Spawner: ".concat(spawner.getClass().getName())); try { // Folia - profiler
-            spawner.tick(this, spawnEnemies);
+                spawner.tick(this, spawnEnemies);
             } finally { profiler.stopTimer(customSpawnerTimer); } // Folia - profiler
         }
     }
@@ -1102,11 +1081,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         profiler.push("iceandsnow");
 
         if (!this.paperConfig().environment.disableIceAndSnow) { // Paper - Option to disable ice and snow
-        for (int i = 0; i < tickSpeed; i++) {
-            if (simpleRandom.nextInt(48) == 0) { // Paper - optimise random ticking
-                this.tickPrecipitation(this.getBlockRandomPos(minX, 0, minZ, 15));
+            for (int i = 0; i < tickSpeed; i++) {
+                if (simpleRandom.nextInt(48) == 0) { // Paper - optimise random ticking
+                    this.tickPrecipitation(this.getBlockRandomPos(minX, 0, minZ, 15));
+                }
             }
-        }
         } // Paper - Option to disable ice and snow
 
         profiler.popPush("tickBlocks");
@@ -1129,8 +1108,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             if (this.isRainingAt(pos)) {
                 DifficultyInstance difficulty = this.getCurrentDifficultyAt(pos);
                 boolean isTrap = this.getGameRules().get(GameRules.SPAWN_MOBS)
-                    && this.random.nextDouble() < difficulty.getEffectiveDifficulty() * this.paperConfig().entities.spawning.skeletonHorseThunderSpawnChance.or(0.01) // Paper - Configurable spawn chances for skeleton horses
-                    && !this.getBlockState(pos.below()).is(BlockTags.LIGHTNING_RODS);
+                        && this.random.nextDouble() < difficulty.getEffectiveDifficulty() * this.paperConfig().entities.spawning.skeletonHorseThunderSpawnChance.or(0.01) // Paper - Configurable spawn chances for skeleton horses
+                        && !this.getBlockState(pos.below()).is(BlockTags.LIGHTNING_RODS);
                 if (isTrap) {
                     SkeletonHorse horse = EntityTypes.SKELETON_HORSE.create(this, EntitySpawnReason.EVENT);
                     if (horse != null) {
@@ -1188,13 +1167,13 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     public Optional<BlockPos> findLightningRod(final BlockPos center) {
         Optional<BlockPos> nearbyLightningRod = this.getPoiManager()
-            .findClosest(
-                p -> p.is(PoiTypes.LIGHTNING_ROD),
-                lightningRodPos -> lightningRodPos.getY() == this.getHeight(Heightmap.Types.WORLD_SURFACE, lightningRodPos.getX(), lightningRodPos.getZ()) - 1,
-                center,
-                128,
-                PoiManager.Occupancy.ANY
-            );
+                .findClosest(
+                        p -> p.is(PoiTypes.LIGHTNING_ROD),
+                        lightningRodPos -> lightningRodPos.getY() == this.getHeight(Heightmap.Types.WORLD_SURFACE, lightningRodPos.getX(), lightningRodPos.getZ()) - 1,
+                        center,
+                        128,
+                        PoiManager.Occupancy.ANY
+                );
         return nearbyLightningRod.map(blockPos -> blockPos.above(1));
     }
 
@@ -1479,39 +1458,39 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         try {
             // Folia - region threading
             // Paper end - log detailed entity tick information
-        entity.setOldPosAndRot();
-        ProfilerFiller profiler = Profiler.get();
-        entity.tickCount++;
-        entity.totalEntityAge++; // Paper - age-like counter for all entities
-        profiler.push(entity.typeHolder()::getRegisteredName);
-        profiler.incrementCounter("tickNonPassenger");
-        final boolean isActive = io.papermc.paper.entity.activation.ActivationRange.checkIfActive(entity); // Paper - EAR 2
-        // Folia start - profiler
-        final int timerId = isActive ? entity.getType().tickTimerId : entity.getType().inactiveTickTimerId;
-        final ca.spottedleaf.leafprofiler.RegionizedProfiler.Handle foliaProfiler = io.papermc.paper.threadedregions.TickRegionScheduler.getProfiler();
-        foliaProfiler.startTimer(timerId);
-        try {
-        // Folia end - profiler
-        if (isActive) { // Paper - EAR 2
-        entity.tick();
-        // Folia start - region threading
-        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
-            // removed from region while ticking
-            return;
-        }
-        if (entity.handlePortal()) {
-            // portalled
-            return;
-        }
-        // Folia end - region threading
-        } else {entity.inactiveTick();} // Paper - EAR 2
-        profiler.pop();
-        } finally { foliaProfiler.stopTimer(timerId); } // Folia - profiler
+            entity.setOldPosAndRot();
+            ProfilerFiller profiler = Profiler.get();
+            entity.tickCount++;
+            entity.totalEntityAge++; // Paper - age-like counter for all entities
+            profiler.push(entity.typeHolder()::getRegisteredName);
+            profiler.incrementCounter("tickNonPassenger");
+            final boolean isActive = io.papermc.paper.entity.activation.ActivationRange.checkIfActive(entity); // Paper - EAR 2
+            // Folia start - profiler
+            final int timerId = isActive ? entity.getType().tickTimerId : entity.getType().inactiveTickTimerId;
+            final ca.spottedleaf.leafprofiler.RegionizedProfiler.Handle foliaProfiler = io.papermc.paper.threadedregions.TickRegionScheduler.getProfiler();
+            foliaProfiler.startTimer(timerId);
+            try {
+                // Folia end - profiler
+                if (isActive) { // Paper - EAR 2
+                    entity.tick();
+                    // Folia start - region threading
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                        // removed from region while ticking
+                        return;
+                    }
+                    if (entity.handlePortal()) {
+                        // portalled
+                        return;
+                    }
+                    // Folia end - region threading
+                } else {entity.inactiveTick();} // Paper - EAR 2
+                profiler.pop();
+            } finally { foliaProfiler.stopTimer(timerId); } // Folia - profiler
 
-        for (Entity passenger : entity.getPassengers()) {
-            this.tickPassenger(entity, passenger, isActive); // Paper - EAR 2
-        }
-        // Paper start - log detailed entity tick information
+            for (Entity passenger : entity.getPassengers()) {
+                this.tickPassenger(entity, passenger, isActive); // Paper - EAR 2
+            }
+            // Paper start - log detailed entity tick information
         } finally {
             // Folia - region threading
         }
@@ -1528,37 +1507,37 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             foliaProfiler.startTimer(timerId);
             try {
                 // Folia end - profiler
-            entity.setOldPosAndRot();
-            entity.tickCount++;
-            entity.totalEntityAge++; // Paper - age-like counter for all entities
-            ProfilerFiller profiler = Profiler.get();
-            profiler.push(entity.typeHolder()::getRegisteredName);
-            profiler.incrementCounter("tickPassenger");
-            // Paper start - EAR 2
-            if (isActive) {
-            entity.rideTick();
-            // Folia start - region threading
-            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
-                // removed from region while ticking
-                return;
-            }
-            if (entity.handlePortal()) {
-                // portalled
-                return;
-            }
-            // Folia end - region threading
-            } else {
-                entity.setDeltaMovement(Vec3.ZERO);
-                entity.inactiveTick();
-                // copied from inside of if (isPassenger()) of passengerTick, but that ifPassenger is unnecessary
-                vehicle.positionRider(entity);
-            }
-            // Paper end - EAR 2
-            profiler.pop();
+                entity.setOldPosAndRot();
+                entity.tickCount++;
+                entity.totalEntityAge++; // Paper - age-like counter for all entities
+                ProfilerFiller profiler = Profiler.get();
+                profiler.push(entity.typeHolder()::getRegisteredName);
+                profiler.incrementCounter("tickPassenger");
+                // Paper start - EAR 2
+                if (isActive) {
+                    entity.rideTick();
+                    // Folia start - region threading
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                        // removed from region while ticking
+                        return;
+                    }
+                    if (entity.handlePortal()) {
+                        // portalled
+                        return;
+                    }
+                    // Folia end - region threading
+                } else {
+                    entity.setDeltaMovement(Vec3.ZERO);
+                    entity.inactiveTick();
+                    // copied from inside of if (isPassenger()) of passengerTick, but that ifPassenger is unnecessary
+                    vehicle.positionRider(entity);
+                }
+                // Paper end - EAR 2
+                profiler.pop();
 
-            for (Entity passenger : entity.getPassengers()) {
-                this.tickPassenger(entity, passenger, isActive); // Paper - EAR 2
-            }
+                for (Entity passenger : entity.getPassengers()) {
+                    this.tickPassenger(entity, passenger, isActive); // Paper - EAR 2
+                }
             } finally { foliaProfiler.stopTimer(timerId); } // Folia - profiler
         }
     }
@@ -1644,7 +1623,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
 
     public <T extends Entity> void getEntities(
-        final EntityTypeTest<Entity, T> type, final Predicate<? super T> selector, final List<? super T> result, final int maxResults
+            final EntityTypeTest<Entity, T> type, final Predicate<? super T> selector, final List<? super T> result, final int maxResults
     ) {
         this.getEntities().get(type, entity -> {
             if (selector.test(entity)) {
@@ -1760,6 +1739,17 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     // CraftBukkit start
     private boolean addEntity(final Entity entity, final org.bukkit.event.entity.CreatureSpawnEvent.@Nullable SpawnReason spawnReason) {
         org.spigotmc.AsyncCatcher.catchOp("entity add"); // Spigot
+
+        // Florium start - hard optimize
+        if (FloriumEntityScheduler.isLagEntity(entity)) {
+            ChunkPos pos = entity.chunkPosition();
+            ChunkAccess acc = getChunkIfLoadedImmediately(pos.x(), pos.z());
+            if (acc != null) {
+                acc.florium$lagEntities++;
+            }
+        }
+        // Florium end - hard optimize
+
         entity.generation = false; // Paper - Don't fire sync event during generation; Reset flag if it was added during a ServerLevel generation process
         // Paper start - extra debug info
         if (entity.valid) {
@@ -1858,7 +1848,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             float progressFloat = Mth.clamp(progress, 0, 10) / 10.0f;
             org.bukkit.craftbukkit.block.CraftBlock bukkitBlock = org.bukkit.craftbukkit.block.CraftBlock.at(this, blockPos);
             new io.papermc.paper.event.block.BlockBreakProgressUpdateEvent(bukkitBlock, progressFloat, entity.getBukkitEntity())
-                .callEvent();
+                    .callEvent();
         }
         // Paper end - Add BlockBreakProgressUpdateEvent
         for (ServerPlayer player : this.server.getPlayerList().getPlayers()) {
@@ -1880,50 +1870,50 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     @Override
     public void playSeededSound(
-        final @Nullable Entity except,
-        final double x,
-        final double y,
-        final double z,
-        final Holder<SoundEvent> sound,
-        final SoundSource source,
-        final float volume,
-        final float pitch,
-        final long seed
+            final @Nullable Entity except,
+            final double x,
+            final double y,
+            final double z,
+            final Holder<SoundEvent> sound,
+            final SoundSource source,
+            final float volume,
+            final float pitch,
+            final long seed
     ) {
         this.server
-            .getPlayerList()
-            .broadcast(
-                except instanceof Player player ? player : null,
-                x,
-                y,
-                z,
-                sound.value().getRange(volume),
-                this.dimension(),
-                new ClientboundSoundPacket(sound, source, x, y, z, volume, pitch, seed)
-            );
+                .getPlayerList()
+                .broadcast(
+                        except instanceof Player player ? player : null,
+                        x,
+                        y,
+                        z,
+                        sound.value().getRange(volume),
+                        this.dimension(),
+                        new ClientboundSoundPacket(sound, source, x, y, z, volume, pitch, seed)
+                );
     }
 
     @Override
     public void playSeededSound(
-        final @Nullable Entity except,
-        final Entity sourceEntity,
-        final Holder<SoundEvent> sound,
-        final SoundSource source,
-        final float volume,
-        final float pitch,
-        final long seed
+            final @Nullable Entity except,
+            final Entity sourceEntity,
+            final Holder<SoundEvent> sound,
+            final SoundSource source,
+            final float volume,
+            final float pitch,
+            final long seed
     ) {
         this.server
-            .getPlayerList()
-            .broadcast(
-                except instanceof Player player ? player : null,
-                sourceEntity.getX(),
-                sourceEntity.getY(),
-                sourceEntity.getZ(),
-                sound.value().getRange(volume),
-                this.dimension(),
-                new ClientboundSoundEntityPacket(sound, source, sourceEntity, volume, pitch, seed)
-            );
+                .getPlayerList()
+                .broadcast(
+                        except instanceof Player player ? player : null,
+                        sourceEntity.getX(),
+                        sourceEntity.getY(),
+                        sourceEntity.getZ(),
+                        sound.value().getRange(volume),
+                        this.dimension(),
+                        new ClientboundSoundEntityPacket(sound, source, sourceEntity, volume, pitch, seed)
+                );
     }
 
     @Override
@@ -1953,16 +1943,16 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     @Override
     public void levelEvent(final @Nullable Entity source, final int type, final BlockPos pos, final int data) {
         this.server
-            .getPlayerList()
-            .broadcast(
-                source instanceof Player player ? player : null,
-                pos.getX(),
-                pos.getY(),
-                pos.getZ(),
-                64.0, // Paper - diff on change (the 64.0 distance is used as defaults for sound ranges in spigot config for ender dragon, end portal and wither)
-                this.dimension(),
-                new ClientboundLevelEventPacket(type, pos, data, false)
-            );
+                .getPlayerList()
+                .broadcast(
+                        source instanceof Player player ? player : null,
+                        pos.getX(),
+                        pos.getY(),
+                        pos.getZ(),
+                        64.0, // Paper - diff on change (the 64.0 distance is used as defaults for sound ranges in spigot config for ender dragon, end portal and wither)
+                        this.dimension(),
+                        new ClientboundLevelEventPacket(type, pos, data, false)
+                );
     }
 
     public int getLogicalHeight() {
@@ -1990,39 +1980,39 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.getChunkSource().blockChanged(pos);
         regionizedWorldData.pathTypesByPosCache.invalidate(pos); // Folia - region threading
         if (this.paperConfig().misc.updatePathfindingOnBlockUpdate) { // Paper - option to disable pathfinding updates
-        VoxelShape oldShape = old.getCollisionShape(this, pos);
-        VoxelShape newShape = current.getCollisionShape(this, pos);
-        if (Shapes.joinIsNotEmpty(oldShape, newShape, BooleanOp.NOT_SAME)) {
-            List<PathNavigation> navigationsToUpdate = new ObjectArrayList<>();
-
-            try { // Paper - catch CME see below why
-                for (java.util.Iterator<Mob> iterator = regionizedWorldData.getNavigatingMobs(); iterator.hasNext();) { // Folia - region threading
-                Mob navigatingMob = iterator.next(); // Folia - region threading
-                PathNavigation pathNavigation = navigatingMob.getNavigation();
-                if (pathNavigation.shouldRecomputePath(pos)) {
-                    navigationsToUpdate.add(pathNavigation);
+            VoxelShape oldShape = old.getCollisionShape(this, pos);
+            VoxelShape newShape = current.getCollisionShape(this, pos);
+            if (Shapes.joinIsNotEmpty(oldShape, newShape, BooleanOp.NOT_SAME)) {
+                List<PathNavigation> navigationsToUpdate = new ObjectArrayList<>();
+
+                try { // Paper - catch CME see below why
+                    for (java.util.Iterator<Mob> iterator = regionizedWorldData.getNavigatingMobs(); iterator.hasNext();) { // Folia - region threading
+                        Mob navigatingMob = iterator.next(); // Folia - region threading
+                        PathNavigation pathNavigation = navigatingMob.getNavigation();
+                        if (pathNavigation.shouldRecomputePath(pos)) {
+                            navigationsToUpdate.add(pathNavigation);
+                        }
+                    }
+                    // Paper start - catch CME see below why
+                } catch (final java.util.ConcurrentModificationException concurrentModificationException) {
+                    // This can happen because the pathfinder update below may trigger a chunk load, which in turn may cause more navigators to register
+                    // In this case we just run the update again across all the iterators as the chunk will then be loaded
+                    // As this is a relative edge case it is much faster than copying navigators (on either read or write)
+                    this.sendBlockUpdated(pos, old, current, updateFlags);
+                    return;
                 }
-            }
-            // Paper start - catch CME see below why
-            } catch (final java.util.ConcurrentModificationException concurrentModificationException) {
-                // This can happen because the pathfinder update below may trigger a chunk load, which in turn may cause more navigators to register
-                // In this case we just run the update again across all the iterators as the chunk will then be loaded
-                // As this is a relative edge case it is much faster than copying navigators (on either read or write)
-                this.sendBlockUpdated(pos, old, current, updateFlags);
-                return;
-            }
-            // Paper end - catch CME see below why
+                // Paper end - catch CME see below why
 
-            try {
-                //this.isUpdatingNavigations = true; // Folia - region threading
+                try {
+                    //this.isUpdatingNavigations = true; // Folia - region threading
 
-                for (PathNavigation navigation : navigationsToUpdate) {
-                    navigation.recomputePath();
+                    for (PathNavigation navigation : navigationsToUpdate) {
+                        navigation.recomputePath();
+                    }
+                } finally {
+                    //this.isUpdatingNavigations = false; // Folia - region threading
                 }
-            } finally {
-                //this.isUpdatingNavigations = false; // Folia - region threading
             }
-        }
         } // Paper - option to disable pathfinding updates
     }
 
@@ -2045,7 +2035,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     @Override
     public void updateNeighborsAtExceptFromFacing(
-        final BlockPos pos, final Block blockObject, final Direction skipDirection, final @Nullable Orientation orientation
+            final BlockPos pos, final Block blockObject, final Direction skipDirection, final @Nullable Orientation orientation
     ) {
         this.getCurrentWorldData().neighborUpdater.updateNeighborsAtExceptFromFacing(pos, blockObject, skipDirection, orientation); // Folia - region threading
     }
@@ -2057,7 +2047,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     @Override
     public void neighborChanged(
-        final BlockState state, final BlockPos pos, final Block changedBlock, final @Nullable Orientation orientation, final boolean movedByPiston
+            final BlockState state, final BlockPos pos, final Block changedBlock, final @Nullable Orientation orientation, final boolean movedByPiston
     ) {
         this.getCurrentWorldData().neighborUpdater.neighborChanged(state, pos, changedBlock, orientation, movedByPiston); // Folia - region threading
     }
@@ -2079,64 +2069,64 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     @Override
     public void explode(
-        final @Nullable Entity source,
-        final @Nullable DamageSource damageSource,
-        final @Nullable ExplosionDamageCalculator damageCalculator,
-        final double x,
-        final double y,
-        final double z,
-        final float r,
-        final boolean fire,
-        final Level.ExplosionInteraction interactionType,
-        final ParticleOptions smallExplosionParticles,
-        final ParticleOptions largeExplosionParticles,
-        final WeightedList<ExplosionParticleInfo> blockParticles,
-        final Holder<SoundEvent> explosionSound
+            final @Nullable Entity source,
+            final @Nullable DamageSource damageSource,
+            final @Nullable ExplosionDamageCalculator damageCalculator,
+            final double x,
+            final double y,
+            final double z,
+            final float r,
+            final boolean fire,
+            final Level.ExplosionInteraction interactionType,
+            final ParticleOptions smallExplosionParticles,
+            final ParticleOptions largeExplosionParticles,
+            final WeightedList<ExplosionParticleInfo> blockParticles,
+            final Holder<SoundEvent> explosionSound
     ) {
         // CraftBukkit start
         this.explode0(source, damageSource, damageCalculator, x, y, z, r, fire, interactionType, smallExplosionParticles, largeExplosionParticles, blockParticles, explosionSound);
     }
 
     public ServerExplosion explode0(
-        final @Nullable Entity source,
-        final @Nullable DamageSource damageSource,
-        final @Nullable ExplosionDamageCalculator damageCalculator,
-        final double x,
-        final double y,
-        final double z,
-        final float r,
-        final boolean fire,
-        final Level.ExplosionInteraction interactionType,
-        final ParticleOptions smallExplosionParticles,
-        final ParticleOptions largeExplosionParticles,
-        final WeightedList<ExplosionParticleInfo> blockParticles,
-        final Holder<SoundEvent> explosionSound
+            final @Nullable Entity source,
+            final @Nullable DamageSource damageSource,
+            final @Nullable ExplosionDamageCalculator damageCalculator,
+            final double x,
+            final double y,
+            final double z,
+            final float r,
+            final boolean fire,
+            final Level.ExplosionInteraction interactionType,
+            final ParticleOptions smallExplosionParticles,
+            final ParticleOptions largeExplosionParticles,
+            final WeightedList<ExplosionParticleInfo> blockParticles,
+            final Holder<SoundEvent> explosionSound
     ) {
         return this.explode0(source, damageSource, damageCalculator, x, y, z, r, fire, interactionType, smallExplosionParticles, largeExplosionParticles, blockParticles, explosionSound, null);
     }
     public ServerExplosion explode0(
-        final @Nullable Entity source,
-        final @Nullable DamageSource damageSource,
-        final @Nullable ExplosionDamageCalculator damageCalculator,
-        final double x,
-        final double y,
-        final double z,
-        final float r,
-        final boolean fire,
-        final Level.ExplosionInteraction interactionType,
-        final ParticleOptions smallExplosionParticles,
-        final ParticleOptions largeExplosionParticles,
-        final WeightedList<ExplosionParticleInfo> blockParticles,
-        final Holder<SoundEvent> explosionSound,
-        java.util.function.@Nullable Consumer<ServerExplosion> configurator
+            final @Nullable Entity source,
+            final @Nullable DamageSource damageSource,
+            final @Nullable ExplosionDamageCalculator damageCalculator,
+            final double x,
+            final double y,
+            final double z,
+            final float r,
+            final boolean fire,
+            final Level.ExplosionInteraction interactionType,
+            final ParticleOptions smallExplosionParticles,
+            final ParticleOptions largeExplosionParticles,
+            final WeightedList<ExplosionParticleInfo> blockParticles,
+            final Holder<SoundEvent> explosionSound,
+            java.util.function.@Nullable Consumer<ServerExplosion> configurator
     ) {
         // CraftBukkit end
         Explosion.BlockInteraction blockInteraction = switch (interactionType) {
             case NONE -> Explosion.BlockInteraction.KEEP;
             case BLOCK -> this.getDestroyType(GameRules.BLOCK_EXPLOSION_DROP_DECAY);
             case MOB -> this.getGameRules().get(GameRules.MOB_GRIEFING)
-                ? this.getDestroyType(GameRules.MOB_EXPLOSION_DROP_DECAY)
-                : Explosion.BlockInteraction.KEEP;
+                    ? this.getDestroyType(GameRules.MOB_EXPLOSION_DROP_DECAY)
+                    : Explosion.BlockInteraction.KEEP;
             case TNT -> this.getDestroyType(GameRules.TNT_EXPLOSION_DROP_DECAY);
             case TRIGGER -> Explosion.BlockInteraction.TRIGGER_BLOCK;
             case STANDARD -> Explosion.BlockInteraction.DESTROY; // CraftBukkit - handle custom explosion type
@@ -2182,16 +2172,16 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             if (this.shouldTickBlocksAt(eventData.pos())) {
                 if (this.doBlockEvent(eventData)) {
                     this.server
-                        .getPlayerList()
-                        .broadcast(
-                            null,
-                            eventData.pos().getX(),
-                            eventData.pos().getY(),
-                            eventData.pos().getZ(),
-                            64.0,
-                            this.dimension(),
-                            new ClientboundBlockEventPacket(eventData.pos(), eventData.block(), eventData.paramA(), eventData.paramB())
-                        );
+                            .getPlayerList()
+                            .broadcast(
+                                    null,
+                                    eventData.pos().getX(),
+                                    eventData.pos().getY(),
+                                    eventData.pos().getZ(),
+                                    64.0,
+                                    this.dimension(),
+                                    new ClientboundBlockEventPacket(eventData.pos(), eventData.block(), eventData.paramA(), eventData.paramB())
+                            );
                 }
             } else {
                 blockEventsToReschedule.add(eventData); // Folia - regionised ticking
@@ -2230,69 +2220,69 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
 
     public <T extends ParticleOptions> int sendParticles(
-        final T particle,
-        final double x,
-        final double y,
-        final double z,
-        final int count,
-        final double xDist,
-        final double yDist,
-        final double zDist,
-        final double speed
+            final T particle,
+            final double x,
+            final double y,
+            final double z,
+            final int count,
+            final double xDist,
+            final double yDist,
+            final double zDist,
+            final double speed
     ) {
         return this.sendParticlesSource(null, particle, false, false, x, y, z, count, xDist, yDist, zDist, speed); // CraftBukkit - visibility api support
     }
 
     public <T extends ParticleOptions> int sendParticles(
-        final T particle,
-        final boolean overrideLimiter,
-        final boolean alwaysShow,
-        final double x,
-        final double y,
-        final double z,
-        final int count,
-        final double xDist,
-        final double yDist,
-        final double zDist,
-        final double speed
+            final T particle,
+            final boolean overrideLimiter,
+            final boolean alwaysShow,
+            final double x,
+            final double y,
+            final double z,
+            final int count,
+            final double xDist,
+            final double yDist,
+            final double zDist,
+            final double speed
     ) {
         // Paper start - visibility api support
         return this.sendParticlesSource(null, particle, overrideLimiter, alwaysShow, x, y, z, count, xDist, yDist, zDist, speed);
     }
     public <T extends ParticleOptions> int sendParticlesSource(
-        @Nullable Entity sender,
-        T options,
-        boolean overrideLimiter,
-        boolean alwaysShow,
-        double x,
-        double y,
-        double z,
-        int count,
-        double xDist,
-        double yDist,
-        double zDist,
-        double speed
+            @Nullable Entity sender,
+            T options,
+            boolean overrideLimiter,
+            boolean alwaysShow,
+            double x,
+            double y,
+            double z,
+            int count,
+            double xDist,
+            double yDist,
+            double zDist,
+            double speed
     ) {
         return sendParticlesSource(this.getLocalPlayers(), sender, options, overrideLimiter, alwaysShow, x, y, z, count, xDist, yDist, zDist, speed); // Folia - region threading
     }
     public <T extends ParticleOptions> int sendParticlesSource(
-        List<ServerPlayer> receivers,
-        @Nullable Entity sender,
-        T particle,
-        boolean overrideLimiter,
-        boolean alwaysShow,
-        double x,
-        double y,
-        double z,
-        int count,
-        double xDist,
-        double yDist,
-        double zDist,
-        double speed
+            List<ServerPlayer> receivers,
+            @Nullable Entity sender,
+            T particle,
+            boolean overrideLimiter,
+            boolean alwaysShow,
+            double x,
+            double y,
+            double z,
+            int count,
+            double xDist,
+            double yDist,
+            double zDist,
+            double speed
     ) {
         // Paper end - visibility api support
         ClientboundLevelParticlesPacket packet = new ClientboundLevelParticlesPacket(
-            particle, overrideLimiter, alwaysShow, x, y, z, (float)xDist, (float)yDist, (float)zDist, (float)speed, count
+                particle, overrideLimiter, alwaysShow, x, y, z, (float)xDist, (float)yDist, (float)zDist, (float)speed, count
         );
         int result = 0;
 
@@ -2308,27 +2298,27 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
 
     public <T extends ParticleOptions> boolean sendParticles(
-        final ServerPlayer player,
-        final T particle,
-        final boolean overrideLimiter,
-        final boolean alwaysShow,
-        final double x,
-        final double y,
-        final double z,
-        final int count,
-        final double xDist,
-        final double yDist,
-        final double zDist,
-        final double speed
+            final ServerPlayer player,
+            final T particle,
+            final boolean overrideLimiter,
+            final boolean alwaysShow,
+            final double x,
+            final double y,
+            final double z,
+            final int count,
+            final double xDist,
+            final double yDist,
+            final double zDist,
+            final double speed
     ) {
         Packet<?> packet = new ClientboundLevelParticlesPacket(
-            particle, overrideLimiter, alwaysShow, x, y, z, (float)xDist, (float)yDist, (float)zDist, (float)speed, count
+                particle, overrideLimiter, alwaysShow, x, y, z, (float)xDist, (float)yDist, (float)zDist, (float)speed, count
         );
         return this.sendParticles(player, overrideLimiter, x, y, z, packet);
     }
 
     private boolean sendParticles(
-        final ServerPlayer player, final boolean overrideLimiter, final double x, final double y, final double z, final Packet<?> packet
+            final ServerPlayer player, final boolean overrideLimiter, final double x, final double y, final double z, final Packet<?> packet
     ) {
         if (player.level() != this) {
             return false;
@@ -2384,7 +2374,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
 
     public @Nullable BlockPos findNearestMapStructure(
-        final TagKey<Structure> structureTag, final BlockPos origin, final int maxSearchRadius, final boolean createReference
+            final TagKey<Structure> structureTag, final BlockPos origin, final int maxSearchRadius, final boolean createReference
     ) {
         Optional<HolderSet.Named<Structure>> tag = this.registryAccess().lookupOrThrow(Registries.STRUCTURE).get(structureTag);
         if (tag.isEmpty()) {
@@ -2392,24 +2382,24 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         }
 
         Pair<BlockPos, Holder<Structure>> result = this.getChunkSource()
-            .getGenerator()
-            .findNearestMapStructure(this, tag.get(), origin, maxSearchRadius, createReference);
+                .getGenerator()
+                .findNearestMapStructure(this, tag.get(), origin, maxSearchRadius, createReference);
         return result != null ? result.getFirst() : null;
     }
 
     public @Nullable Pair<BlockPos, Holder<Biome>> findClosestBiome3d(
-        final Predicate<Holder<Biome>> biomeTest,
-        final BlockPos origin,
-        final int maxSearchRadius,
-        final int sampleResolutionHorizontal,
-        final int sampleResolutionVertical
+            final Predicate<Holder<Biome>> biomeTest,
+            final BlockPos origin,
+            final int maxSearchRadius,
+            final int sampleResolutionHorizontal,
+            final int sampleResolutionVertical
     ) {
         return this.getChunkSource()
-            .getGenerator()
-            .getBiomeSource()
-            .findClosestBiome3d(
-                origin, maxSearchRadius, sampleResolutionHorizontal, sampleResolutionVertical, biomeTest, this.getChunkSource().randomState().sampler(), this
-            );
+                .getGenerator()
+                .getBiomeSource()
+                .findClosestBiome3d(
+                        origin, maxSearchRadius, sampleResolutionHorizontal, sampleResolutionVertical, biomeTest, this.getChunkSource().randomState().sampler(), this
+                );
     }
 
     // Folia start - cache world border
@@ -2649,28 +2639,28 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     private static void dumpEntities(final Writer output, final Iterable<Entity> entities) throws IOException {
         CsvOutput csvOutput = CsvOutput.builder()
-            .addColumn("x")
-            .addColumn("y")
-            .addColumn("z")
-            .addColumn("uuid")
-            .addColumn("type")
-            .addColumn("alive")
-            .addColumn("display_name")
-            .addColumn("custom_name")
-            .build(output);
+                .addColumn("x")
+                .addColumn("y")
+                .addColumn("z")
+                .addColumn("uuid")
+                .addColumn("type")
+                .addColumn("alive")
+                .addColumn("display_name")
+                .addColumn("custom_name")
+                .build(output);
 
         for (Entity entity : entities) {
             Component customName = entity.getCustomName();
             Component displayName = entity.getDisplayName();
             csvOutput.writeRow(
-                entity.getX(),
-                entity.getY(),
-                entity.getZ(),
-                entity.getUUID(),
-                entity.typeHolder().getRegisteredName(),
-                entity.isAlive(),
-                displayName.getString(),
-                customName != null ? customName.getString() : null
+                    entity.getX(),
+                    entity.getY(),
+                    entity.getZ(),
+                    entity.getUUID(),
+                    entity.typeHolder().getRegisteredName(),
+                    entity.isAlive(),
+                    displayName.getString(),
+                    customName != null ? customName.getString() : null
             );
         }
     }
@@ -2723,16 +2713,16 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     @VisibleForTesting
     public String getWatchdogStats() {
         return String.format(
-            Locale.ROOT,
-            "players: %s, entities: %s [%s], block_entities: %d [%s], block_ticks: %d, fluid_ticks: %d, chunk_source: %s",
-            this.players.size(),
-            this.moonrise$getEntityLookup().getDebugInfo(), // Paper - rewrite chunk system
-            getTypeCount(this.moonrise$getEntityLookup().getAll(), e -> e.typeHolder().getRegisteredName()), // Paper - rewrite chunk system
-            0, // Folia - region threading
-            "null", // Folia - region threading
-            this.getBlockTicks().count(),
-            this.getFluidTicks().count(),
-            this.gatherChunkSourceStats()
+                Locale.ROOT,
+                "players: %s, entities: %s [%s], block_entities: %d [%s], block_ticks: %d, fluid_ticks: %d, chunk_source: %s",
+                this.players.size(),
+                this.moonrise$getEntityLookup().getDebugInfo(), // Paper - rewrite chunk system
+                getTypeCount(this.moonrise$getEntityLookup().getAll(), e -> e.typeHolder().getRegisteredName()), // Paper - rewrite chunk system
+                0, // Folia - region threading
+                "null", // Folia - region threading
+                this.getBlockTicks().count(),
+                this.getFluidTicks().count(),
+                this.gatherChunkSourceStats()
         );
     }
 
@@ -2747,11 +2737,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
             Comparator<Entry<String>> compareByCount = Comparator.comparingInt(Entry::getIntValue);
             return countByType.object2IntEntrySet()
-                .stream()
-                .sorted(compareByCount.reversed())
-                .limit(5L)
-                .map(ex -> (String)ex.getKey() + ":" + ex.getIntValue())
-                .collect(Collectors.joining(","));
+                    .stream()
+                    .sorted(compareByCount.reversed())
+                    .limit(5L)
+                    .map(ex -> (String)ex.getKey() + ":" + ex.getIntValue())
+                    .collect(Collectors.joining(","));
         } catch (Exception e) {
             return "";
         }
@@ -2916,15 +2906,15 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         WeatherData weatherData = this.getWeatherData();
         category.setDetail("Loaded entity count", () -> String.valueOf(this.moonrise$getEntityLookup().getEntityCount())); // Paper - rewrite chunk system
         category.setDetail(
-            "Server weather",
-            () -> String.format(
-                Locale.ROOT,
-                "Rain time: %d (now: %b), thunder time: %d (now: %b)",
-                weatherData.getRainTime(),
-                this.isRaining(),
-                weatherData.getThunderTime(),
-                this.isThundering()
-            )
+                "Server weather",
+                () -> String.format(
+                        Locale.ROOT,
+                        "Rain time: %d (now: %b), thunder time: %d (now: %b)",
+                        weatherData.getRainTime(),
+                        this.isRaining(),
+                        weatherData.getThunderTime(),
+                        this.isThundering()
+                )
         );
         return category;
     }
@@ -3023,7 +3013,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                 if (false && ServerLevel.this.isUpdatingNavigations) { // Paper - Remove unnecessary onTrackingStart during navigation warning
                     String message = "onTrackingStart called during navigation iteration";
                     Util.logAndPauseIfInIde(
-                        "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
+                            "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
                     );
                 }
 
@@ -3065,10 +3055,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                         }
 
                         synchronized (map) { // Folia - make map data thread-safe
-                        map.carriedByPlayers.remove(player);
-                        if (map.carriedBy.removeIf(holdingPlayer -> holdingPlayer.player == player)) {
-                            map.decorations.remove(player.getName().getString());
-                        }
+                            map.carriedByPlayers.remove(player);
+                            if (map.carriedBy.removeIf(holdingPlayer -> holdingPlayer.player == player)) {
+                                map.decorations.remove(player.getName().getString());
+                            }
                         } // Folia - make map data thread-safe
                     }
                 }
@@ -3097,7 +3087,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                 if (false && ServerLevel.this.isUpdatingNavigations) { // Paper - Remove unnecessary onTrackingStart during navigation warning
                     String message = "onTrackingStart called during navigation iteration";
                     Util.logAndPauseIfInIde(
-                        "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
+                            "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
                     );
                 }
 
@@ -3122,6 +3112,16 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             }
             // CraftBukkit end
             new com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent(entity.getBukkitEntity(), ServerLevel.this.getWorld()).callEvent(); // Paper - fire while valid
+
+            // Florium start - hard optimize
+            if (FloriumEntityScheduler.isLagEntity(entity)) {
+                ChunkPos pos = entity.chunkPosition();
+                ChunkAccess acc = getChunkIfLoadedImmediately(pos.x(), pos.z());
+                if (acc != null) {
+                    acc.florium$lagEntities--;
+                }
+            }
+            // Florium end - hard optimize
         }
 
         @Override
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index 480b6b2405706fd4d7158d0e160adafd84099887..33287e67306edf88d36c8bb9b36592998173c017 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -74,6 +74,10 @@ public class FallingBlockEntity extends Entity {
     protected static final EntityDataAccessor<BlockPos> DATA_START_POS = SynchedEntityData.defineId(FallingBlockEntity.class, EntityDataSerializers.BLOCK_POS);
     public boolean autoExpire = true; // Paper - Expand FallingBlock API
 
+    // Florium start - hard optimization
+    public int florium$skippedTicks = 0;
+    // Florium end
+
     public FallingBlockEntity(final EntityType<? extends FallingBlockEntity> type, final Level level) {
         super(type, level);
     }
diff --git a/net/minecraft/world/level/chunk/ChunkAccess.java b/net/minecraft/world/level/chunk/ChunkAccess.java
index 28f703204afd834cd50335346ef065670d6b37ad..6ecb1e1e11fd97a59ed63ede12c41f24e9423e97 100644
--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -90,6 +90,10 @@ public abstract class ChunkAccess implements LightChunk, StructureAccess, BiomeM
     private volatile boolean[] skyEmptinessMap;
     private volatile boolean[] blockEmptinessMap;
 
+    // Florium start - hard optimize
+    public volatile int florium$lagEntities = 0;
+    // Florium end
+
     @Override
     public ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] starlight$getBlockNibbles() {
         return this.blockNibbles;
